package/base-files/files/lib/upgrade/nand.sh

```diff
@@ -242,55 +241,131 @@ nand_upgrade_ubifs() {
+	fi
+
+	if ! nand_upgrade_tar_dir_has_assets "$tar_file" "$board_dir" ; then
+		>&2 echo "No assets found for '${board_dir}/', falling back to legacy method."
+		board_dir=$(tar t -f "$tar_file" | grep -m 1 '^sysupgrade-.*/$')
+		board_dir=${board_dir%/}
+		if ! nand_upgrade_tar_dir_has_assets "$tar_file" "$board_dir" ; then
+			nand_do_upgrade_abort "No assets found for '${board_dir}/'."
+		fi
+	fi
 
-	local kernel_length=`(tar xf $tar_file ${board_dir}/kernel -O | wc -c) 2> /dev/null`
-	local rootfs_length=`(tar xf $tar_file ${board_dir}/root -O | wc -c) 2> /dev/null`
+	>&2 echo "Assets from '${board_dir}/'."
 
-	local rootfs_type="$(identify_tar "$tar_file" ${board_dir}/root)"
+	local kernel_length=$( (tar xf "$tar_file" "${board_dir}/kernel" -O | wc -c) 2> /dev/null )
+	local rootfs_length=$( (tar xf "$tar_file" "${board_dir}/root" -O | wc -c) 2> /dev/null )
 
-	local has_kernel=1
-	local has_env=0
+	local rootfs_type="$(identify_tar "$tar_file" "${board_dir}/root")"
 
-	[ "$kernel_length" != 0 -a -n "$kernel_mtd" ] && {
+	local has_ubi_kernel=1
+	local has_ubi_env=0
+
+	[ "$kernel_length" = 0 ] || [ -n "$kernel_mtd" ] && has_ubi_kernel=0
+
+	nand_upgrade_prepare_ubi "$rootfs_length" "$rootfs_type" "$has_ubi_kernel" "$has_ubi_env"
+	err=$?
+	[ "$err" != 0 ] && nand_do_upgrade_abort "nand_upgrade_prepare_ubi returned '$err'"
+
+	if [ "$kernel_length" != 0 ] && [ -n "$kernel_mtd" ] ; then
 		tar xf $tar_file ${board_dir}/kernel -O | mtd write - $CI_KERNPART
-	}
-	[ "$kernel_length" = 0 -o ! -z "$kernel_mtd" ] && has_kernel=0
+	fi
 
-	nand_upgrade_prepare_ubi "$rootfs_length" "$rootfs_type" "$has_kernel" "$has_env"
+	local ubidev="$(nand_find_ubi $CI_UBIPART)"
 
-	local ubidev="$( nand_find_ubi "$CI_UBIPART" )"
-	[ "$has_kernel" = "1" ] && {
+	if [ "$has_ubi_kernel" = "1" ] ; then
 		local kern_ubivol="$(nand_find_volume $ubidev $CI_KERNPART)"
 		tar xf $tar_file ${board_dir}/kernel -O | \
 			ubiupdatevol /dev/$kern_ubivol -s $kernel_length -
-	}
+	fi
 
 	local root_ubivol="$(nand_find_volume $ubidev $CI_ROOTPART)"
+
 	tar xf $tar_file ${board_dir}/root -O | \
 		ubiupdatevol /dev/$root_ubivol -s $rootfs_length -
 
 	nand_do_upgrade_success
 }
 
 # Recognize type of passed file and start the upgrade process
+
 nand_do_upgrade() {
 	local file_type=$(identify $1)
 
 	[ ! "$(find_mtd_index "$CI_UBIPART")" ] && CI_UBIPART="rootfs"
 
 	case "$file_type" in
-		"ubi")		nand_upgrade_ubinized $1;;
-		"ubifs")	nand_upgrade_ubifs $1;;
-		*)		nand_upgrade_tar $1;;
+		"ubi")		nand_upgrade_ubinized "$1";;
+		"ubifs")	nand_upgrade_ubifs "$1";;
+		*)		nand_upgrade_tar "$1";;
 	esac
 }
 
+nand_do_upgrade_success() {
+	local conf_tar="/tmp/sysupgrade.tgz"
+
+	sync
+	[ -f "$conf_tar" ] && nand_restore_config "$conf_tar"
+	echo "sysupgrade completed"
+	umount -a
+	reboot -f
+}
+
+nand_do_upgrade_abort() {
+	>&2 echo "##### UPGRADE ABORTED: $1 #####"
+	>&2 echo "Rebooting in 5 seconds"
+	sleep 5
+	sync
+	umount -a
+	reboot -f
+}
+
+
+
+
```

Comment:
Any reason for four empty lines?

Reply:
My recollection was to clearly separate "current" code from "legacy" code.

Reply:
But there is no need at all to do such a separate with 4 lines. It doesn't make a better separation than one line I think.
checkpatch.pl in kernel tree always tell you to remove more than one extra empty lines in source code files.